pkg/controller/core.oam.dev/v1alpha2/envbinding/binding.go

```diff
@@ -0,0 +1,284 @@
+/*
+ Copyright 2021. The KubeVela Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package envbinding
+
+import (
+	"context"
+
+	"github.com/ghodss/yaml"
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/apis/types"
+	"github.com/oam-dev/kubevela/pkg/appfile"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+// EnvBindApp describes the app bound to the environment
+type EnvBindApp struct {
+	baseApp   *v1beta1.Application
+	newApp    *v1beta1.Application
+	envConfig *v1beta1.EnvConfig
+
+	componentManifests []*types.ComponentManifest
+	assembledManifests []*unstructured.Unstructured
+
+	ManifestWork *unstructured.Unstructured
+}
+
+// NewEnvBindApp create EnvBindApp
+func NewEnvBindApp(base *v1beta1.Application, envConfig *v1beta1.EnvConfig) *EnvBindApp {
+	return &EnvBindApp{
+		baseApp:   base,
+		envConfig: envConfig,
+	}
+}
+
+// GenerateConfiguredApplication patch component parameters to base Application
+func (e *EnvBindApp) GenerateConfiguredApplication() error {
+	newApp := e.baseApp.DeepCopy()
+	var baseComponent *v1beta1.ApplicationComponent
+	var matchIdx int
+	for patchIdx := range e.envConfig.Patch.Components {
+		patchComponent := e.envConfig.Patch.Components[patchIdx]
+		var IsMatched bool
+		for baseIdx := range e.baseApp.Spec.Components {
+			component := e.baseApp.Spec.Components[baseIdx]
+			if patchComponent.Name == component.Name && patchComponent.Type == component.Type {
+				matchIdx, baseComponent = baseIdx, &component
+				IsMatched = true
+				break
+			}
+		}
+		if !IsMatched || baseComponent == nil {
+			return errors.Errorf("fail to match component %s", patchComponent.Name)
+		}
+		targetComponent, err := PatchComponent(baseComponent, &patchComponent)
+		if err != nil {
+			return err
+		}
+		newApp.Spec.Components[matchIdx] = *targetComponent
+	}
+	e.newApp = newApp
+	return nil
+}
+
+// Render render the configured Application
+func (e *EnvBindApp) Render(ctx context.Context, appParser *appfile.Parser) error {
```

Comment:
```suggestion
func (e *EnvBindApp) render(ctx context.Context, appParser *appfile.Parser) error {
```

It seems that this function will only be called inside the package.